### PR TITLE
Using a GPT model fine-tuned with a contrastive loss

### PR DESCRIPTION
**Pull request by LouisCastricato**
Currently we cannot scale CLOOB past declutr-base for unknown reasons. Current hypothesis is that we need to pretrain or fine-tune a language model using a contrastive loss. Kharr and I are exploring using [SimCTG](https://arxiv.org/abs/2202.06417) as a method to fine-tune a base model before training a CARP CLOOB model.

### CODE REVIEW

**Comment by LouisCastricato** [comment]
This PR makes mixed precision more intelligent, moves around a lot of the configuration stuff, adds more options to the configuration, adds the ability to use custom tokenizers, and adds experiments for using GPT as the base encoder

**Comment by LouisCastricato** [comment]
Going to merge this as it has a number of bug fixes for magiCARP as a whole.
